File: timeline/test/interval.test.ts
import { Inclusivity, Interval } from '../src/interval';
import { Temporal } from '@js-temporal/polyfill';
import './jest.setup';
import Duration = Temporal.Duration;

const dur = (m: number) => Duration.from({ minutes: m });
const between = Interval.Between;
const at = Interval.At;

test(Inclusivity.opposite.name, () => {
  expect(Inclusivity.opposite(Inclusivity.Inclusive)).toEqual(Inclusivity.Exclusive);
});

test(Inclusivity.compareRestrictiveness.name, () => {
  expect(Inclusivity.compareRestrictiveness(Inclusivity.Inclusive, Inclusivity.Exclusive)).toEqual(-1);
  expect(Inclusivity.compareRestrictiveness(Inclusivity.Exclusive, Inclusivity.Exclusive)).toEqual(0);
});

test(Interval.At.name, () => {
  const time = dur(45);

  const i = at(time);
  expect(i.start).toEqual(time);
  expect(i.end).toEqual(time);
  expect(i.startInclusivity).toEqual(Inclusivity.Inclusive);
  expect(i.endInclusivity).toEqual(Inclusivity.Inclusive);
});

test(Interval.Between.name, () => {
  const t1 = dur(1);
  const t2 = dur(5);

  let i = between(t1, t2);
  expect(i.start).toEqual(t1);
  expect(i.end).toEqual(t2);
  expect(i.startInclusivity).toEqual(Inclusivity.Inclusive);
  expect(i.endInclusivity).toEqual(Inclusivity.Inclusive);

  i = between(t1, t2, Inclusivity.Exclusive);
  expect(i.start).toEqual(t1);
  expect(i.end).toEqual(t2);
  expect(i.startInclusivity).toEqual(Inclusivity.Exclusive);
  expect(i.endInclusivity).toEqual(Inclusivity.Exclusive);

  i = between(t1, t2, Inclusivity.Inclusive, Inclusivity.Exclusive);
  expect(i.start).toEqual(t1);
  expect(i.end).toEqual(t2);
  expect(i.startInclusivity).toEqual(Inclusivity.Inclusive);
  expect(i.endInclusivity).toEqual(Inclusivity.Exclusive);
});

test(Interval.prototype.isEmpty.name, () => {
  expect(at(dur(5)).isEmpty()).toBeFalsy();
  expect(between(dur(1), dur(2)).isEmpty()).toBeFalsy();
  expect(between(dur(2), dur(1)).isEmpty()).toBeTruthy();
  expect(between(dur(1), dur(1), Inclusivity.Inclusive, Inclusivity.Exclusive).isEmpty()).toBeTruthy();
});

test(Interval.prototype.isSingleton.name, () => {
  expect(at(dur(1)).isSingleton()).toBeTruthy();
  expect(between(dur(1), dur(2)).isSingleton()).toBeFalsy();
  expect(between(dur(2), dur(1)).isSingleton()).toBeFalsy();
});

test(Interval.prototype.duration.name, () => {
  expect(between(dur(1), dur(2)).duration()).toEqual(dur(1));
});

test(Interval.intersect.name, () => {
  expect(Interval.intersect(between(dur(1), dur(2)), between(dur(0), dur(4)))).toEqual(between(dur(1), dur(2)));
  expect(Interval.intersect(between(dur(0), dur(2)), between(dur(1), dur(3)))).toEqual(between(dur(1), dur(2)));
  expect(Interval.intersect(between(dur(0), dur(2), Inclusivity.Exclusive), between(dur(1), dur(3)))).toEqual(
    between(dur(1), dur(2), Inclusivity.Inclusive, Inclusivity.Exclusive)
  );
  expect(Interval.intersect(between(dur(0), dur(1)), between(dur(3), dur(4))).isEmpty()).toBeTruthy();
});

test(Interval.union.name, () => {
  expect(Interval.union(between(dur(0), dur(2)), at(dur(1)))).toEqual(between(dur(0), dur(2)));
  expect(Interval.union(between(dur(0), dur(1)), between(dur(1), dur(2)))).toEqual(between(dur(0), dur(2)));
  expect(Interval.union(between(dur(0), dur(1), Inclusivity.Exclusive), between(dur(1), dur(2)))).toEqual(
    between(dur(0), dur(2), Inclusivity.Exclusive, Inclusivity.Inclusive)
  );
  expect(Interval.union(between(dur(0), dur(1)), at(dur(4)))).toBeUndefined();
});

test(Interval.subtract.name, () => {
  expect(Interval.subtract(between(dur(0), dur(2)), at(dur(1)))).toEqual([
    between(dur(0), dur(1), Inclusivity.Inclusive, Inclusivity.Exclusive),
    between(dur(1), dur(2), Inclusivity.Exclusive, Inclusivity.Inclusive)
  ]);
  expect(Interval.subtract(between(dur(0), dur(1)), between(dur(2), dur(3)))).toEqual([between(dur(0), dur(1))]);
});

test(Interval.compareStarts.name, () => {
  expect(Interval.compareStarts(between(dur(0), dur(1)), at(dur(1)))).toEqual(-1);
  expect(Interval.compareStarts(between(dur(0), dur(1), Inclusivity.Exclusive), at(dur(0)))).toEqual(1);
  expect(Interval.compareStarts(between(dur(0), dur(1)), at(dur(0)))).toEqual(0);
});

test(Interval.compareEnds.name, () => {
  expect(Interval.compareEnds(between(dur(0), dur(1)), at(dur(0)))).toEqual(1);
  expect(Interval.compareEnds(between(dur(0), dur(1), Inclusivity.Exclusive), at(dur(1)))).toEqual(-1);
  expect(Interval.compareEnds(between(dur(0), dur(1)), at(dur(1)))).toEqual(0);
});

test(Interval.compareEndToStart.name, () => {
  expect(Interval.compareEndToStart(at(dur(0)), at(dur(1)))).toEqual(-1);
  expect(Interval.compareEndToStart(at(dur(0)), between(dur(0), dur(1), Inclusivity.Exclusive))).toEqual(0);
  expect(Interval.compareEndToStart(between(dur(0), dur(2)), between(dur(1), dur(4)))).toEqual(1);
});

test(Interval.prototype.contains.name, () => {
  expect(between(dur(0), dur(2)).contains(dur(1))).toBeTruthy();
  expect(between(dur(0), dur(2)).contains(dur(3))).toBeFalsy();
  expect(between(dur(0), dur(2)).contains(at(dur(1)))).toBeTruthy();
  expect(between(dur(0), dur(2)).contains(at(dur(3)))).toBeFalsy();
  expect(between(dur(0), dur(2)).contains(between(dur(1), dur(3)))).toBeFalsy();
  expect(between(dur(0), dur(2)).contains(between(dur(1), dur(2)))).toBeTruthy();
});

test(Interval.prototype.shiftBy.name, () => {
  expect(between(dur(0), dur(1)).shiftBy(dur(1))).toEqual(between(dur(1), dur(2)));
  expect(at(dur(2)).shiftBy(dur(-1))).toEqual(at(dur(1)));
  expect(between(dur(0), dur(4)).shiftBy(dur(1), dur(-1))).toEqual(between(dur(1), dur(3)));
  expect(between(dur(0), dur(2)).shiftBy(dur(1), dur(-1))).toEqual(at(dur(1)));
  expect(between(dur(0), dur(1)).shiftBy(dur(2), dur(0)).isEmpty()).toBeTruthy();
});

File: procedural/timeline/src/test/kotlin/gov/nasa/ammos/aerie/timeline/IntervalTest.kt
package gov.nasa.ammos.aerie.timeline

import gov.nasa.jpl.aerie.merlin.protocol.types.Duration
import gov.nasa.jpl.aerie.merlin.protocol.types.Duration.seconds
import gov.nasa.ammos.aerie.timeline.Interval.Companion.EMPTY
import gov.nasa.ammos.aerie.timeline.Interval.Companion.at
import gov.nasa.ammos.aerie.timeline.Interval.Companion.between
import gov.nasa.ammos.aerie.timeline.Interval.Companion.betweenClosedOpen
import gov.nasa.ammos.aerie.timeline.Interval.Inclusivity.*
import gov.nasa.ammos.aerie.timeline.util.duration.rangeTo
import gov.nasa.ammos.aerie.timeline.util.duration.rangeUntil
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class IntervalTest {
  @Nested
  inner class InclusivityTest {
    @Test
    fun opposite() {
      assertEquals(Inclusive, Exclusive.opposite())
    }

    @Test
    fun moreRestrictiveThan() {
      assertTrue(Exclusive.moreRestrictiveThan(Inclusive))
      assertFalse(Exclusive.moreRestrictiveThan(Exclusive))
    }
  }

  @Test
  fun at() {
    val t = seconds(45)

    val i = at(t)
    assertEquals(t, i.start)
    assertEquals(t, i.end)
    assertEquals(Inclusive, i.startInclusivity)
    assertEquals(Inclusive, i.endInclusivity)
  }

  @Test
  fun between() {
    val t1 = seconds(1)
    val t2 = seconds(5)

    var i = between(t1, t2)
    assertEquals(t1, i.start)
    assertEquals(t2, i.end)
    assertEquals(Inclusive, i.startInclusivity)
    assertEquals(Inclusive, i.endInclusivity)

    i = between(t1, t2, Exclusive)
    assertEquals(Exclusive, i.startInclusivity)
    assertEquals(Exclusive, i.endInclusivity)

    i = between(t1, t2, Inclusive, Exclusive)
    assertEquals(Inclusive, i.startInclusivity)
    assertEquals(Exclusive, i.endInclusivity)

    val i2 = betweenClosedOpen(t1, t2)
    assertEquals(i, i2)
  }

  @Test
  fun isEmpty() {
    assertFalse(at(seconds(5)).isEmpty())
    assertFalse((seconds(1) .. seconds(2)).isEmpty())
    assertTrue((seconds(2) .. seconds(1)).isEmpty())
    assertTrue(between(seconds(1), seconds(1), Inclusive, Exclusive).isEmpty())
    assertTrue(EMPTY.isEmpty())
  }

  @Test
  fun isSingleton() {
    assertTrue(at(seconds(1)).isPoint())
    assertFalse((seconds(1) .. seconds(2)).isPoint())
    assertFalse((seconds(2) .. seconds(1)).isPoint())
  }

  @Test
  fun duration() {
    assertEquals(seconds(1), (seconds(1) .. seconds(2)).duration())
    assertEquals(Duration.ZERO, EMPTY.duration())
    assertEquals(Duration.ZERO, (seconds(2) .. seconds(1)).duration())
  }

  @Test
  fun intersect() {
    assertEquals(
        (seconds(1) .. seconds(2)),
        seconds(1) .. seconds(2) intersection seconds(0) .. seconds(4)
    )
    assertEquals(
        (seconds(1) .. seconds(2)),
        (seconds(0) .. seconds(2)).intersection(seconds(1) .. seconds(3))
    )
    assertEquals(
        betweenClosedOpen(seconds(1), seconds(2)),
        between(seconds(0), seconds(2), Exclusive).intersection(seconds(1) .. seconds(3))
    )
    assertTrue((seconds(0) .. seconds(1)).intersection(seconds(2) .. seconds(3)).isEmpty())
    assertTrue((seconds(0) .. seconds(1)).intersection(EMPTY).isEmpty())
  }

  @Test
  fun union() {
    assertIterableEquals(
        listOf(seconds(0) .. seconds(2)),
        seconds(0) .. seconds(2) union at(seconds(1))
    )
    assertIterableEquals(
        listOf(seconds(0) .. seconds(2)),
        (seconds(0) .. seconds(2)).union(EMPTY)
    )
    assertIterableEquals(
        listOf(seconds(0) .. seconds(2)),
        (seconds(0) .. seconds(1)).union(seconds(1) .. seconds(2))
    )
    assertIterableEquals(
        listOf(seconds(0) .. seconds(2)),
        betweenClosedOpen(seconds(0), seconds(1)).union(seconds(1) .. seconds(2))
    )
    assertIterableEquals(
        listOf((seconds(0) .. seconds(1)), at(seconds(4))),
        (seconds(0) .. seconds(1)).union(at(seconds(4)))
    )
    assertIterableEquals(
        listOf(seconds(0) .. seconds(1)),
        (seconds(0) .. seconds(1)).union(EMPTY)
    )
  }

  @Test
  fun subtract() {
    assertIterableEquals(
        listOf(betweenClosedOpen(seconds(0), seconds(1)), between(seconds(1), seconds(2), Exclusive, Inclusive)),
        (seconds(0) .. seconds(2)).minus(at(seconds(1)))
    )
    assertIterableEquals(
        listOf(seconds(0) .. seconds(1)),
        (seconds(0) .. seconds(1)).minus(seconds(3) .. seconds(4))
    )
    assertIterableEquals(
        listOf(betweenClosedOpen(seconds(0), seconds(1))),
        (seconds(0) .. seconds(2)).minus(seconds(1) .. seconds(4))
    )
    assertIterableEquals(
        listOf(seconds(3) .. seconds(4)),
        (seconds(3) .. seconds(4)).minus(seconds(1) .. seconds(2))
    )
    assertIterableEquals(
        listOf(between(seconds(3), seconds(4), Exclusive, Inclusive)),
        (seconds(2) .. seconds(4)).minus(seconds(1) .. seconds(3))
    )
  }

  @Test
  fun compareStarts() {
    assertEquals(-1, (seconds(0) .. seconds(1)).compareStarts(at(seconds(1))))
    assertEquals(1, between(seconds(0), seconds(1), Exclusive).compareStarts(at(seconds(0))))
    assertEquals(0, (seconds(0) .. seconds(1)).compareStarts(at(seconds(0))))
  }

  @Test
  fun compareEnds() {
    assertEquals(1, (seconds(0) .. seconds(1)).compareEnds(at(seconds(0))))
    assertEquals(-1, between(seconds(0), seconds(1), Exclusive).compareEnds(at(seconds(1))))
    assertEquals(0, (seconds(0) .. seconds(1)).compareEnds(at(seconds(1))))
    assertEquals(-1, (seconds(0) ..< seconds(1)).compareEnds(at(seconds(1))))
  }

  @Test
  fun compareEndToStart() {
    assertEquals(-1, at(seconds(0)).compareEndToStart(at(seconds(1))))
    assertEquals(0, at(seconds(0)).compareEndToStart(between(seconds(0), seconds(1), Exclusive)))
    assertEquals(1, (seconds(0) .. seconds(2)).compareEndToStart(seconds(1) .. seconds(4)))
  }

  @Test
  fun contains() {
    assertTrue((seconds(0) .. seconds(2)).contains(seconds(1)))
    assertFalse((seconds(0) .. seconds(2)).contains(seconds(3)))
    assertTrue((seconds(0) .. seconds(2)).contains(at(seconds(1))))
    assertFalse((seconds(0) .. seconds(2)).contains(at(seconds(3))))
    assertTrue((seconds(0) .. seconds(2)).contains(seconds(1) .. seconds(2)))
    assertFalse((seconds(0) .. seconds(2)).contains(seconds(1) .. seconds(3)))
  }

  @Test
  fun shiftBy() {
    assertEquals((seconds(1) .. seconds(2)), (seconds(0) .. seconds(1)).shiftBy(seconds(1)))
    assertEquals(at(seconds(1)), at(seconds(2)).shiftBy(seconds(-1)))
    assertEquals((seconds(1) .. seconds(3)), (seconds(0) .. seconds(4)).shiftBy(seconds(1), seconds(-1)))
    assertEquals(at(seconds(1)), (seconds(0) .. seconds(2)).shiftBy(seconds(1), seconds(-1)))
    assertTrue((seconds(0) .. seconds(1)).shiftBy(seconds(2), seconds(0)).isEmpty())
  }
}
